### PR TITLE
feat: toggle_floating_state and is_floating (#306)

### PR DESCRIPTION
**Pull request by codybloemhard**
In order to resolve issue #306

Adds:

- StackSet<Clone + PartialEq + Eq + Hash>::is_floating
- StackSet<Xid>::toggle_floating_state
- builtin::actions::floating::toggle_floating_focused

### CODE REVIEW

File: src/pure/stack_set.rs [hunk 1]
```diff
@@ -799,6 +804,26 @@ impl StackSet<Xid> {
...
+    /// not currently mapped to a screen. This is required to determine the correct
+    /// relative positioning for the floating client as is it is moved between
+    /// screens.
+    pub fn toggle_floating_state(&mut self, client: Xid, r: Rect) -> Result<Option<Rect>> {
+        Ok(if self.is_floating(&client) {
+            self.sink(&client)
+        } else {
+            self.float(client, r)?;
+            None
+        })
```
Comment (sminez) [inline src/pure/stack_set.rs]: Can this be bound to a variable and returned as `Ok(rect)` rather wrapping the whole if statement

**Review by sminez — CHANGES_REQUESTED** [review]
Implementation looks good 👍 

If you can add some teats for the toggle method then this should be good to merge.There are equivalent tests for similar methods in the file already using the test_case macro to cover the different cases you highlight in the doc comment.

**Comment by codybloemhard** [comment]
I tried to split `toggle_floating_state` test case into two, one using `test_case` to check for the different errors it can return but I couldn't assert them because `Error` has no `PartialEq` and quickly adding it into the derive didn't work out.

**Review by sminez — APPROVED** [review]
A couple of minor style things but I'll fix those up post merge rather than nitpicking on this PR :slightly_smiling_face: 

Thank you for taking the time to contribute to Penrose!

**Pull request merged**